File: app/src/main/java/com/itn/securebrowser/util/BlockDataStore.java
package com.itn.securebrowser.util;

import android.content.Context;
import android.content.SharedPreferences;
import org.json.JSONArray;
import org.json.JSONObject;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class BlockDataStore {

    private static final String KEY_GROUPS = "block_groups";
    private static final String KEY_SITES = "block_sites";
    private final SharedPreferences prefs;

    public BlockDataStore(Context context) {
        prefs = context.getSharedPreferences("itn_block_data", Context.MODE_PRIVATE);
    }

    // --- Data classes ---

    public static class BlockSchedule {
        public final List<String> days;
        public final String from;
        public final String to;

        public BlockSchedule(List<String> days, String from, String to) {
            this.days = days;
            this.from = from;
            this.to = to;
        }
    }

    public static class BlockGroup {
        public final String name;
        public final List<String> domains;
        public final HashMap<String, Integer> dailyLimits;
        public final List<BlockSchedule> schedules;

        public BlockGroup(String name, List<String> domains, HashMap<String, Integer> dailyLimits, List<BlockSchedule> schedules) {
            this.name = name;
            this.domains = domains;
            this.dailyLimits = dailyLimits;
            this.schedules = schedules;
        }
    }

    public static class BlockSite {
        public final String domain;
        public final HashMap<String, Integer> dailyLimits;
        public final List<BlockSchedule> schedules;

        public BlockSite(String domain, HashMap<String, Integer> dailyLimits, List<BlockSchedule> schedules) {
            this.domain = domain;
            this.dailyLimits = dailyLimits;
            this.schedules = schedules;
        }
    }

    public static abstract class BlockMatch {
        public static class GroupMatch extends BlockMatch {
            public final BlockGroup group;
            public GroupMatch(BlockGroup group) { this.group = group; }
        }
        public static class SiteMatch extends BlockMatch {
            public final BlockSite site;
            public SiteMatch(BlockSite site) { this.site = site; }
        }
    }

    // --- Groups ---

    public void saveGroup(BlockGroup group) {
        for (BlockSite s : getSites()) {
            if (group.domains.contains(s.domain)) {
                throw new IllegalStateException("Domain '" + s.domain + "' is an individual site. Remove it first.");
            }
        }
        List<BlockGroup> list = getGroups();
        int idx = -1;
        for (int i = 0; i < list.size(); i++) {
            if (list.get(i).name.equals(group.name)) { idx = i; break; }
        }
        if (idx >= 0) list.set(idx, group); else list.add(group);
        prefs.edit().putString(KEY_GROUPS, serializeGroups(list)).apply();
    }

    public List<BlockGroup> getGroups() {
        String json = prefs.getString(KEY_GROUPS, null);
        if (json == null) return new ArrayList<>();
        try {
            JSONArray arr = new JSONArray(json);
            List<BlockGroup> list = new ArrayList<>();
            for (int i = 0; i < arr.length(); i++) {
                list.add(deserializeGroup(arr.getJSONObject(i)));
            }
            return list;
        } catch (Exception e) {
            return new ArrayList<>();
        }
    }

    public void deleteGroup(String name) {
        List<BlockGroup> list = new ArrayList<>();
        for (BlockGroup g : getGroups()) {
            if (!g.name.equals(name)) list.add(g);
        }
        prefs.edit().putString(KEY_GROUPS, serializeGroups(list)).apply();
    }

    // --- Sites ---

    public void saveSite(BlockSite site) {
        for (BlockGroup g : getGroups()) {
            if (g.domains.contains(site.domain)) {
                throw new IllegalStateException("Domain '" + site.domain + "' is in group '" + g.name + "'.");
            }
        }
        List<BlockSite> list = getSites();
        int idx = -1;
        for (int i = 0; i < list.size(); i++) {
            if (list.get(i).domain.equals(site.domain)) { idx = i; break; }
        }
        if (idx >= 0) list.set(idx, site); else list.add(site);
        prefs.edit().putString(KEY_SITES, serializeSites(list)).apply();
    }

    public List<BlockSite> getSites() {
        String json = prefs.getString(KEY_SITES, null);
        if (json == null) return new ArrayList<>();
        try {
            JSONArray arr = new JSONArray(json);
            List<BlockSite> list = new ArrayList<>();
            for (int i = 0; i < arr.length(); i++) {
                list.add(deserializeSite(arr.getJSONObject(i)));
            }
            return list;
        } catch (Exception e) {
            return new ArrayList<>();
        }
    }

    public void deleteSite(String domain) {
        List<BlockSite> list = new ArrayList<>();
        for (BlockSite s : getSites()) {
            if (!s.domain.equals(domain)) list.add(s);
        }
        prefs.edit().putString(KEY_SITES, serializeSites(list)).apply();
    }

    // --- Lookup ---

    public BlockMatch findDomain(String domain) {
        for (BlockGroup g : getGroups()) {
            if (g.domains.contains(domain)) return new BlockMatch.GroupMatch(g);
        }
        for (BlockSite s : getSites()) {
            if (s.domain.equals(domain)) return new BlockMatch.SiteMatch(s);
        }
        String[] parts = domain.split("\\.");
        for (int i = 1; i < parts.length - 1; i++) {
            StringBuilder parent = new StringBuilder();
            for (int j = i; j < parts.length; j++) {
                if (j > i) parent.append(".");
                parent.append(parts[j]);
            }
            String p = parent.toString();
            for (BlockGroup g : getGroups()) {
                if (g.domains.contains(p)) return new BlockMatch.GroupMatch(g);
            }
            for (BlockSite s : getSites()) {
                if (s.domain.equals(p)) return new BlockMatch.SiteMatch(s);
            }
        }
        return null;
    }

    // --- Clear ---

    public void clearAll() {
        prefs.edit().clear().apply();
    }

    // --- JSON serialization ---

    private JSONObject serializeSchedule(BlockSchedule s) throws Exception {
        JSONObject o = new JSONObject();
        o.put("days", new JSONArray(s.days));
        o.put("from", s.from);
        o.put("to", s.to);
        return o;
    }

    private JSONObject serializeLimits(HashMap<String, Integer> limits) throws Exception {
        JSONObject o = new JSONObject();
        for (HashMap.Entry<String, Integer> e : limits.entrySet()) {
            o.put(e.getKey(), e.getValue());
        }
        return o;
    }

    private String serializeGroups(List<BlockGroup> list) {
        try {
            JSONArray arr = new JSONArray();
            for (BlockGroup g : list) {
                JSONObject o = new JSONObject();
                o.put("name", g.name);
                o.put("domains", new JSONArray(g.domains));
                o.put("dailyLimits", serializeLimits(g.dailyLimits));
                JSONArray schedArr = new JSONArray();
                for (BlockSchedule s : g.schedules) schedArr.put(serializeSchedule(s));
                o.put("schedules", schedArr);
                arr.put(o);
            }
            return arr.toString();
        } catch (Exception e) {
            return "[]";
        }
    }

    private String serializeSites(List<BlockSite> list) {
        try {
            JSONArray arr = new JSONArray();
            for (BlockSite s : list) {
                JSONObject o = new JSONObject();
                o.put("domain", s.domain);
                o.put("dailyLimits", serializeLimits(s.dailyLimits));
                JSONArray schedArr = new JSONArray();
                for (BlockSchedule sc : s.schedules) schedArr.put(serializeSchedule(sc));
                o.put("schedules", schedArr);
                arr.put(o);
            }
            return arr.toString();
        } catch (Exception e) {
            return "[]";
        }
    }

    private BlockSchedule deserializeSchedule(JSONObject o) throws Exception {
        JSONArray daysArr = o.getJSONArray("days");
        List<String> days = new ArrayList<>();
        for (int i = 0; i < daysArr.length(); i++) days.add(daysArr.getString(i));
        return new BlockSchedule(days, o.getString("from"), o.getString("to"));
    }

    private HashMap<String, Integer> deserializeLimits(JSONObject o) throws Exception {
        HashMap<String, Integer> map = new HashMap<>();
        java.util.Iterator<String> keys = o.keys();
        while (keys.hasNext()) {
            String k = keys.next();
            map.put(k, o.getInt(k));
        }
        return map;
    }

    private BlockGroup deserializeGroup(JSONObject o) throws Exception {
        JSONArray domainsArr = o.getJSONArray("domains");
        List<String> domains = new ArrayList<>();
        for (int i = 0; i < domainsArr.length(); i++) domains.add(domainsArr.getString(i));
        JSONArray schedArr = o.getJSONArray("schedules");
        List<BlockSchedule> schedules = new ArrayList<>();
        for (int i = 0; i < schedArr.length(); i++) schedules.add(deserializeSchedule(schedArr.getJSONObject(i)));
        return new BlockGroup(o.getString("name"), domains, deserializeLimits(o.getJSONObject("dailyLimits")), schedules);
    }

    private BlockSite deserializeSite(JSONObject o) throws Exception {
        JSONArray schedArr = o.getJSONArray("schedules");
        List<BlockSchedule> schedules = new ArrayList<>();
        for (int i = 0; i < schedArr.length(); i++) schedules.add(deserializeSchedule(schedArr.getJSONObject(i)));
        return new BlockSite(o.getString("domain"), deserializeLimits(o.getJSONObject("dailyLimits")), schedules);
    }
}

File: app/src/main/java/com/itn/securebrowser/BlockDataStore.kt
package com.itn.securebrowser

import android.content.Context
import android.content.SharedPreferences
import org.json.JSONArray
import org.json.JSONObject

// ══════════════════════════════════════════════════════════════════════════════
// Data classes
// ══════════════════════════════════════════════════════════════════════════════

/**
 * نافذة حظر زمنية.
 * @param days  أسماء الأيام بالإنجليزية الكبيرة: "SATURDAY" … "FRIDAY"
 * @param from  وقت بداية الحظر بصيغة "HH:mm"
 * @param to    وقت نهاية الحظر بصيغة "HH:mm"  (يمكن أن يعبر منتصف الليل)
 */
data class BlockSchedule(
    val days: List<String>,
    val from: String,
    val to: String
)

/**
 * مجموعة مواقع بحدود ورا جداول مشتركة.
 * @param name        اسم المجموعة (مفتاح فريد)
 * @param domains     قائمة النطاقات (بدون www.)
 * @param dailyLimits يوم → دقائق مسموح بها. الأيام الغائبة = بلا حد.
 * @param schedules   نوافذ الحظر الزمنية
 */
data class BlockGroup(
    val name: String,
    val domains: List<String>,
    val dailyLimits: Map<String, Int>,
    val schedules: List<BlockSchedule>
)

/**
 * موقع منفرد خارج أي مجموعة.
 * @param domain      النطاق (بدون www.) — مفتاح فريد
 * @param dailyLimits يوم → دقائق مسموح بها. الأيام الغائبة = بلا حد.
 * @param schedules   نوافذ الحظر الزمنية
 */
data class BlockSite(
    val domain: String,
    val dailyLimits: Map<String, Int>,
    val schedules: List<BlockSchedule>
)

/**
 * نتيجة البحث عن نطاق:
 * - [GroupMatch] إذا كان النطاق ضمن مجموعة
 * - [SiteMatch]  إذا كان موقعاً منفرداً
 */
sealed class BlockMatch {
    data class GroupMatch(val group: BlockGroup) : BlockMatch()
    data class SiteMatch(val site: BlockSite) : BlockMatch()
}

// ══════════════════════════════════════════════════════════════════════════════
// BlockDataStore
// ══════════════════════════════════════════════════════════════════════════════

class BlockDataStore(context: Context) {

    private val prefs: SharedPreferences =
        context.getSharedPreferences("itn_block_data", Context.MODE_PRIVATE)

    companion object {
        private const val KEY_GROUPS = "block_groups"
        private const val KEY_SITES  = "block_sites"
    }

    // ── Groups ────────────────────────────────────────────────────────────────

    /**
     * يحفظ مجموعة (إنشاء أو تعديل حسب الاسم).
     * @throws IllegalStateException إذا كان أحد نطاقات المجموعة مسجّلاً كموقع منفرد.
     */
    fun saveGroup(group: BlockGroup) {
        val conflictSite = getSites().firstOrNull { it.domain in group.domains }
        if (conflictSite != null) {
            throw IllegalStateException(
                "النطاق '${conflictSite.domain}' مسجّل كموقع منفرد — " +
                "أزله أولاً قبل إضافته إلى مجموعة."
            )
        }
        val list = getGroups().toMutableList()
        val idx  = list.indexOfFirst { it.name == group.name }
        if (idx >= 0) list[idx] = group else list.add(group)
        prefs.edit().putString(KEY_GROUPS, serializeGroups(list)).apply()
    }

    /** يُرجع قائمة كل المجموعات المحفوظة. */
    fun getGroups(): List<BlockGroup> {
        val json = prefs.getString(KEY_GROUPS, null) ?: return emptyList()
        return try {
            val arr = JSONArray(json)
            List(arr.length()) { i -> deserializeGroup(arr.getJSONObject(i)) }
        } catch (e: Exception) {
            emptyList()
        }
    }

    /** يحذف المجموعة بالاسم (لا يفعل شيئاً إن لم تُوجد). */
    fun deleteGroup(name: String) {
        val list = getGroups().filter { it.name != name }
        prefs.edit().putString(KEY_GROUPS, serializeGroups(list)).apply()
    }

    // ── Sites ─────────────────────────────────────────────────────────────────

    /**
     * يحفظ موقعاً منفرداً (إنشاء أو تعديل حسب النطاق).
     * @throws IllegalStateException إذا كان النطاق موجوداً داخل مجموعة.
     */
    fun saveSite(site: BlockSite) {
        val conflictGroup = getGroups().firstOrNull { site.domain in it.domains }
        if (conflictGroup != null) {
            throw IllegalStateException(
                "النطاق '${site.domain}' موجود داخل مجموعة '${conflictGroup.name}' — " +
                "لا يمكن إضافته كموقع منفرد."
            )
        }
        val list = getSites().toMutableList()
        val idx  = list.indexOfFirst { it.domain == site.domain }
        if (idx >= 0) list[idx] = site else list.add(site)
        prefs.edit().putString(KEY_SITES, serializeSites(list)).apply()
    }

    /** يُرجع قائمة كل المواقع المنفردة المحفوظة. */
    fun getSites(): List<BlockSite> {
        val json = prefs.getString(KEY_SITES, null) ?: return emptyList()
        return try {
            val arr = JSONArray(json)
            List(arr.length()) { i -> deserializeSite(arr.getJSONObject(i)) }
        } catch (e: Exception) {
            emptyList()
        }
    }

    /** يحذف الموقع المنفرد بالنطاق (لا يفعل شيئاً إن لم يُوجد). */
    fun deleteSite(domain: String) {
        val list = getSites().filter { it.domain != domain }
        prefs.edit().putString(KEY_SITES, serializeSites(list)).apply()
    }

    // ── Lookup ────────────────────────────────────────────────────────────────

    /**
     * يبحث عن [domain] في المجموعات أولاً ثم في المواقع المنفردة.
     * @return [BlockMatch] إذا كان النطاق خاضعاً لقيد، أو null إذا كان حراً.
     */
    fun findDomain(domain: String): BlockMatch? {
        // ① مطابقة تامة
        getGroups().firstOrNull { domain in it.domains }
            ?.let { return BlockMatch.GroupMatch(it) }
        getSites().firstOrNull { it.domain == domain }
            ?.let { return BlockMatch.SiteMatch(it) }

        // ② مطابقة بالنطاق الأب: m.youtube.com → youtube.com
        // يحل مشكلة المتصفحات التي تفتح النسخة المحمولة على نطاق فرعي
        val parts = domain.split(".")
        for (i in 1 until parts.size - 1) {
            val parent = parts.drop(i).joinToString(".")
            getGroups().firstOrNull { parent in it.domains }
                ?.let { return BlockMatch.GroupMatch(it) }
            getSites().firstOrNull { it.domain == parent }
                ?.let { return BlockMatch.SiteMatch(it) }
        }

        return null
    }

    // ══════════════════════════════════════════════════════════════════════════
    // JSON — Serialization
    // ══════════════════════════════════════════════════════════════════════════

    private fun serializeSchedule(s: BlockSchedule): JSONObject = JSONObject().apply {
        put("days", JSONArray().also { arr -> s.days.forEach { arr.put(it) } })
        put("from", s.from)
        put("to",   s.to)
    }

    private fun serializeLimits(limits: Map<String, Int>): JSONObject =
        JSONObject().apply { limits.forEach { (day, mins) -> put(day, mins) } }

    private fun serializeGroup(g: BlockGroup): JSONObject = JSONObject().apply {
        put("name",        g.name)
        put("domains",     JSONArray().also { arr -> g.domains.forEach  { arr.put(it) } })
        put("dailyLimits", serializeLimits(g.dailyLimits))
        put("schedules",   JSONArray().also { arr -> g.schedules.forEach { arr.put(serializeSchedule(it)) } })
    }

    private fun serializeSite(s: BlockSite): JSONObject = JSONObject().apply {
        put("domain",      s.domain)
        put("dailyLimits", serializeLimits(s.dailyLimits))
        put("schedules",   JSONArray().also { arr -> s.schedules.forEach { arr.put(serializeSchedule(it)) } })
    }

    private fun serializeGroups(list: List<BlockGroup>): String =
        JSONArray().also { arr -> list.forEach { arr.put(serializeGroup(it)) } }.toString()

    private fun serializeSites(list: List<BlockSite>): String =
        JSONArray().also { arr -> list.forEach { arr.put(serializeSite(it)) } }.toString()

    // ══════════════════════════════════════════════════════════════════════════
    // JSON — Deserialization
    // ══════════════════════════════════════════════════════════════════════════

    private fun deserializeSchedule(obj: JSONObject): BlockSchedule {
        val daysArr = obj.getJSONArray("days")
        return BlockSchedule(
            days = List(daysArr.length()) { i -> daysArr.getString(i) },
            from = obj.getString("from"),
            to   = obj.getString("to")
        )
    }

    private fun deserializeLimits(obj: JSONObject): Map<String, Int> =
        mutableMapOf<String, Int>().also { map ->
            obj.keys().forEach { key -> map[key] = obj.getInt(key) }
        }

    private fun deserializeSchedules(arr: JSONArray): List<BlockSchedule> =
        List(arr.length()) { i -> deserializeSchedule(arr.getJSONObject(i)) }

    private fun deserializeGroup(obj: JSONObject): BlockGroup {
        val domainsArr = obj.getJSONArray("domains")
        return BlockGroup(
            name        = obj.getString("name"),
            domains     = List(domainsArr.length()) { i -> domainsArr.getString(i) },
            dailyLimits = deserializeLimits(obj.getJSONObject("dailyLimits")),
            schedules   = deserializeSchedules(obj.getJSONArray("schedules"))
        )
    }

    private fun deserializeSite(obj: JSONObject): BlockSite =
        BlockSite(
            domain      = obj.getString("domain"),
            dailyLimits = deserializeLimits(obj.getJSONObject("dailyLimits")),
            schedules   = deserializeSchedules(obj.getJSONArray("schedules"))
        )
}
